File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] output = new int[nums.length];
        for(int i=0;i<nums.length;i++){
            output[i]=1;
        }
        int l=1;
        int r=1;
        for(int i=0;i<nums.length;i++){
            output[i]=output[i]*l;
            l=l*nums[i];
        }
        for(int i=nums.length-1;i>=0;i--){
            output[i]=output[i]*r;
            r=r*nums[i];
        }

        return output;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        r = 1
        l = 1

        for i in range(len(nums)):
            res[i] *= l
            l *= nums[i]

        for i in reversed(range(len(nums))):
            res[i] *= r
            r *= nums[i]

        return res
